Issues/23 voluntary sign up

PangKung2018 opened the pull request:
using twilio API to implement the phone number OTPverification.

Front-end:
![](https://i.imgur.com/1LcYcaC.png)

OTP: (constrain of resending OTP for 30 second)
![](https://i.imgur.com/D7l4Cat.png)

Receive OTP:
![](https://i.imgur.com/IMxJR4V.jpg)

SimonPang2018 committed:
Front-end and backend implement  OTP verification

using twilio API to implement the phone number
 verification.

SimonPang2018 committed:
Merge branch 'master' into issues/23-Voluntary-sign-up

cskeogh commented:
You use real SMSs rather than mock SMSs? How do you propose the Twilio account works? If there a shared account for all Slowvid developers? Or each developer signs up to Twilio? 5c per sms? Do we need the $6 per month package? https://www.twilio.com/sms/pricing/au

PangKung2018 commented:
There is not needed to pay for the Twilio account, there is a $15.5 credit for every new project, I don't know is it because I selected student tier when I create a project, every successful verification spent $0.11, I think it is enough to demonstrate 100times. Also, if the credit is gone, I reckon we can just create a new project in Twilio can copy the key into our app again.

SimonPang2018 committed:
remove the linkless href for Terms of service within signup page

SimonPang2018 committed:
bugs fixes

cskeogh commented on back-end/src/App.ts:
```diff
@@ -1,22 +1,30 @@
-import { Database } from './Model/Database'
-import { UserModel} from './Model/User'
+// import { Database } from './Model/Database'
+// import { UserModel} from './Model/User'
```
Please remove comment code

cskeogh commented on back-end/src/app.js:
```diff
@@ -0,0 +1,16 @@
+const express  = require('express');
+const cors = require('cors');
```
You don't have to use Typescript, however put your code in .ts files not .js for consistency with the project

cskeogh commented on back-end/src/controllers/signupController.js:
```diff
@@ -0,0 +1,6 @@
+ 
+// exports.register = async (req, res) => {
+
+// };
```
What does this file do?

cskeogh commented on front-end/package.json:
```diff
@@ -3,6 +3,8 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "@material-ui/core": "^4.11.0",
+    "@material-ui/icons": "^4.9.1",
```
The group discussed various UI libraries, and decided on Bulma. Please use Bulma.

cskeogh commented on front-end/src/components/Otp.js:
```diff
@@ -0,0 +1,153 @@
+import React, {createRef} from 'react';
+
+export default function Otp({helper, otp, setOtp}) {
```
Components should be in .tsx files

cskeogh commented on front-end/src/components/Signin.js:
```diff
@@ -0,0 +1,239 @@
+            }
+        })
+        .then(data => console.log(data))
+        .catch(err => console.log(err));
```
For asynchronous programming, don't combine await and then. Pick the 'await' method or the 'then' method.

cskeogh commented on front-end/src/components/Signin.js:
```diff
@@ -0,0 +1,239 @@
+                    if(data.valid === true)
+                        alert("This phone number was successfully verified, moving to the home page.");
+                    else
+                        alert("Oops! Wrong OTP!");
```
It always prints "Oops! Wrong OTP" even if code is right

cskeogh commented on front-end/src/components/Signin.js:
```diff
@@ -0,0 +1,239 @@
+            otpShow: false,
+            otp: '',
+            time: {},
+            seconds: 30
```
30 seconds isn't very long to enter the code. I suggest something longer.

cskeogh commented on front-end/src/components/Signin.js:
```diff
@@ -0,0 +1,239 @@
+            //     backgroundColor: 'rgba(160, 160, 160, 0.2)',
+            //     height: '100vh'
+            // }}
+            >
```
Remove commented code

cskeogh commented on front-end/src/components/Signin.js:
```diff
@@ -0,0 +1,239 @@
+                        </div>
+                        {
+                            !this.state.otpShow ?
+                                <p>By tapping Verify an SMS may be sent. Message & data rates may apply.</p> :
+                                null
```
This message is displayed to the Slowvid user "Message & data rates may apply." Are there plans where it costs money to receive SMS? I suggest remove or reword to indicate it's about the receiving of SMS.

cskeogh commented on front-end/src/pages/SignInPage.tsx:
```diff
@@ -1,8 +1,15 @@
+// import React from "react";
+// import { RouteComponentProps } from "@reach/router";
+
+// let Page = (props: RouteComponentProps) => <div>Sign in page goes here</div>;
+
+// export default Page;
+
```
Remove commented code

cskeogh commented on back-end/src/controllers/verifyController.js:
```diff
@@ -0,0 +1,40 @@
+const client = require('twilio')(process.env.ACCOUNT_SID, process.env.AUTH_TOKEN);
+ 
+ 
+exports.getCode = async (req, res) => {
+    console.log("******** getCode");
+    console.log(req.query.phonenumber);
+    console.log("channel: "+req.query.channel);
+    client
+        .verify
+        .services(process.env.VERIFY_SERVICE_SID)
+        .verifications
+        .create({
+            to: `+${req.query.phonenumber}`,
+            channel: req.query.channel
+        })
+        .then(data => {
+            console.log("res of getCode:"),
+            console.log(data),
+            res.status(200).send(data)
+        })
+};
```
This has a security flaw. The back-end should check the request came from Slowvid. Otherwise a malicious user could use `curl` and send many sms to any phone number.

It should be rate limited to mitigate against an attack. Something like: rate limit of 3 requests per Slowvid front-end per day.

cskeogh commented on back-end/src/controllers/verifyController.js:
```diff
@@ -0,0 +1,40 @@
+        .verificationChecks
+        .create({
+            to: `+${req.query.phonenumber}`,
+            code: req.query.code
+        })
```
As the code is sent plain text, this depends on #75. Otherwise a possible man-in-the-middle attack

cskeogh commented on back-end/src/App.ts:
```diff
@@ -1,22 +1,30 @@
+
+require('dotenv').config();
+
+const app = require('../src/app.js');
+
```
Why have an `App.ts` call an `app.js?` Why not just one `App`?

cskeogh commented on back-end/src/controllers/verifyController.js:
```diff
@@ -0,0 +1,40 @@
+ 
+ 
+exports.getCode = async (req, res) => {
+    console.log("******** getCode");
+    console.log(req.query.phonenumber);
+    console.log("channel: "+req.query.channel);
```
Remove `console.log` that are used for debugging

cskeogh commented on front-end/src/components/NameForm.tsx:
```diff
@@ -19,8 +19,15 @@ function NameForm() {
   return (
     <div
-      style={{ padding: "10px", border: "1px solid white", marginTop: "20px" }}
+      style={{ padding: "10px", border: "1px solid white", marginTop: "20px", backgroundColor:"green" }}
     >
+      <div
+        style={{  marginBottom: "100px" }}
+      >
+        <p>
+          You are in safe!
+      </p>
+      </div>
```
This change is not related to #23 Voluntary sign-up

cskeogh commented on back-end/src/controllers/verifyController.js:
```diff
@@ -0,0 +1,40 @@
+};
+ 
+ 
+exports.verifyCode = async (req, res,err) => {
```
This has a security flaw. A malicious user could create a denial of service. I think [twilio limits](https://www.twilio.com/docs/verify/api/rate-limits-and-timeouts) 60 requests in one minute, so a malicious user couldn't brute force crack a verification code, but a single malicious user could use all 60 requests forever. This should be rate limited. The back-end should check the request came from Slowvid - as a mitigation against a malicious user using `curl`.

cskeogh replied:
> The back-end should check the request came from Slowvid - as a mitigation against a malicious user using `curl`.

After researching, restricting the request came from Slowvid wouldn't be possible in our mock app. However a real app would restrict using [SafetyNet mobile attestation API](https://developer.android.com/training/safetynet/attestation).

My rate limit comment is still applicable.

cskeogh commented on back-end/src/.env:
```diff
@@ -0,0 +1,7 @@
+ENV=DEVELOPMENT
```
Is `ENV=DEVELOPMENT` used?

cskeogh reviewed (changes requested):
Thank you for the PR.

What are the .DS_Store binary files? What do they do?
Please resolve conflicts with master.

SimonPang2018 committed:
address various problems

- remove commented code
- increase otp resend time
- fix otp verification alway show
-  "Oops! Wrong OTP"

SimonPang2018 committed:
Merge branch 'master' into issues/23-Voluntary-sign-up

SimonPang2018 committed:
remove app.js
